Gemfile

```diff
@@ -1,6 +1,8 @@
 source "https://rubygems.org"
 
 gem "rails", "5.1.4"
+gem "config"
+gem "bcrypt", "3.1.11"
```

Comment:
gem này để làm gì thế em?

Reply:
1. Gem "config"
Gem "config" giúp quản lý và sử dụng dễ dàng các settings trong project. Tất cả các biến, giá trị cần dùng để cấu hình được khai báo tập trung trong 1 hoặc 1 vài file định dạng YML đơn giản. Khi sử dụng chỉ cần dùng câu lệnh ngắn gọn và giản đơn để gọi lại các thành phần trong file cấu hình. Ngoài ra, gem "config" còn hỗ trợ ERB, kế thừa với cú pháp đơn giản, thuận tiện.
2. Gem "bcrypt"
- Gem "bcrypt" được sử dụng để hash (băm) password cùng lệnh has_secure_password. Khi băm mật khẩu, chúng ta đảm bảo rằng kẻ tấn công không thể đăng nhập được vào site kể cả khi chúng có bản sao của database. Sau khi băm, ta sẽ có password_digest với mục đích tính toán ra một chuỗi mã hóa của chuỗi mật khẩu ban đầu sao cho không thể dịch ngược từ chuỗi đã mã hóa về chuỗi mật khẩu ban đầu (irreversible). Mỗi lần băm sẽ ra kết quả có độ dài cố định, nội dung khác nhau với cùng một chuỗi đầu vào. has_secure_password tự động thêm một phương thức xác thực vào đối tượng mẫu tương ứng. Phương thức này quyết định password nhập vào hợp lệ với người dùng tương ứng hay không bằng cách tính toán "digest" của nó và so sánh kết quả với password_digest trong database